From d29e90ab39300e7baa482c5eb9aada3ce5e8877e Mon Sep 17 00:00:00 2001
From: Yehuda Yitschak <yehuday@marvell.com>
Date: Sun, 28 Jun 2015 13:39:40 +0300
Subject: [PATCH 0230/1240] pcie: add pcie initializations specific to AP-806

This initializations are not part of the driver since
this integration will only exist in AP-806 Z0.

Change-Id: I7c3863dd0aebdfad63cf25aa66d9fa87d1effa51
Signed-off-by: Yehuda Yitschak <yehuday@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/20525
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Hanna Hawa <hannah@marvell.com>
---
 arch/arm/cpu/armv8/armada8k/soc-init.c         | 29 ++++++++++++++++++++++++++
 arch/arm/include/asm/arch-armada8k/regs-base.h |  3 +++
 2 files changed, 32 insertions(+)

diff --git a/arch/arm/cpu/armv8/armada8k/soc-init.c b/arch/arm/cpu/armv8/armada8k/soc-init.c
index cb1d0b7..085cb1b 100644
--- a/arch/arm/cpu/armv8/armada8k/soc-init.c
+++ b/arch/arm/cpu/armv8/armada8k/soc-init.c
@@ -23,6 +23,7 @@
 #include <asm/arch-mvebu/soc.h>
 #include <asm/arch-mvebu/unit-info.h>
 #include <asm/arch-armada8k/armada8k.h>
+#include <asm/arch/regs-base.h>
 
 #define ADDRESS_SHIFT			(20)
 #define MAX_CCU_WINDOWS			(8)
@@ -50,8 +51,36 @@ int soc_get_id(void)
 	return CONFIG_ARMADA_8K_SOC_ID;
 }
 
+#ifdef CONFIG_MVEBU_PCIE
+static void soc_pcie_init(void)
+{
+	u32 reg;
+
+	reg = readl(MVEBU_PCIE_MAC_CTL);
+
+	/* Set PCIe transactions towards A2 as:
+	 * - read allocate
+	 * - write non alocate
+	 * - outer sharable */
+	reg &= ~(0xF << 8);
+	reg |= (0x7 << 8);
+
+	/* Set the Port x4 */
+	reg |= (1 << 14);
+
+	/* Enable PCIe unit */
+	reg = 1;
+
+	writel(reg, MVEBU_PCIE_MAC_CTL);
+}
+#endif
+
 struct mvebu_soc_family *soc_init(void)
 {
+#ifdef CONFIG_MVEBU_PCIE
+	soc_pcie_init();
+#endif
+
 	return &a8k_family_info;
 }
 
diff --git a/arch/arm/include/asm/arch-armada8k/regs-base.h b/arch/arm/include/asm/arch-armada8k/regs-base.h
index 2e1f87b..b345735 100644
--- a/arch/arm/include/asm/arch-armada8k/regs-base.h
+++ b/arch/arm/include/asm/arch-armada8k/regs-base.h
@@ -44,4 +44,7 @@
 
 #define MVEBU_LLC_BASE		(MVEBU_REGS_BASE + 0x8000)
 
+#define MVEBU_RFU_BASE		(MVEBU_REGS_BASE + 0x6F0000)
+#define MVEBU_PCIE_MAC_CTL	(MVEBU_RFU_BASE + 0x200)
+
 #endif	/* _REGS_BASE_H_ */
-- 
1.9.1

